internal/relay/relayer.go

```diff
@@ -75,27 +75,33 @@ func (r *Relayer) Run(ctx context.Context, tasks <-chan neutrontypes.RegisteredQ
 			err       error
 		)
 		select {
-		case query := <-tasks:
-			switch query.QueryType {
-			case string(neutrontypes.InterchainQueryTypeKV):
-				msg := &MessageKV{QueryId: query.Id, KVKeys: query.Keys}
-				err = r.processMessageKV(ctx, msg)
-			case string(neutrontypes.InterchainQueryTypeTX):
-				msg := &MessageTX{QueryId: query.Id, TransactionsFilter: query.TransactionsFilter}
-				err = r.processMessageTX(ctx, msg)
-			default:
-				err = fmt.Errorf("unknown query type: %s", query.QueryType)
-			}
-
-			if err != nil {
-				r.logger.Error("could not process message", zap.Uint64("query_id", queryID), zap.Error(err))
-				neutronmetrics.AddFailedRequest(string(queryType), time.Since(start).Seconds())
-			} else {
-				neutronmetrics.AddSuccessRequest(string(queryType), time.Since(start).Seconds())
-			}
 		case <-ctx.Done():
 			r.logger.Info("Context cancelled, shutting down relayer...")
 			return r.stop()
+		default:
+			select {
+			case <-ctx.Done():
```

Comment:
why do we need the inner ctx.Done branch here and in the submitchecker?

Reply:
go select statement don't have prioritisation, so in the following scenario:

```go
select {
        case <-ctx.Done():
                do_something()
        case <- query := <-tasks:
                do_something_else()
        default:
                do_default_case()
}
```
if we have messages in both cases, i.e. context has been closed and a new query is waiting to be handled, a random one will be chosen, not the one defined first in the select statement. so, if the context is done, there is no need to start query handling and then have a context deadline error as the result — it would be confusing.

this trick with case <-ctx.Done() moved to the higher level and a default case handling other stuff is a workaround for this, it adds priority to the case <-ctx.Done() branch cuz the default will only be chosen if the context is not done.

in the inner branch, we also have two cases: context done and a new query, and here it is fine because we're sure the context is not done, yet, so we can try to receive a new query from the tasks channel. if it's arrived, we can process it. otherwise (if there is no query ready to be processed) we will wait for either a new query from the tasks queue of the context to be done